pkg/publish/tarball.go

```diff
@@ -78,6 +78,12 @@ func (t *tar) Publish(_ context.Context, br build.Result, s string) (name.Refere
 		t.refs[ref] = img
 	}
 
+	log.Printf("Saving %v", t.file)
+	if err := tarball.MultiRefWriteToFile(t.file, t.refs); err != nil {
```

Comment:
This makes writing to a tarball super slow :/

Reply:
This is the same as what was in `Close`, how is this any different?

Reply:
`ko resolve` et al will call `Publish` multiple times, so we would end up serializing and writing the tarball once for every image that gets published instead of just once per invocation.

Reply:
https://www.joelonsoftware.com/2001/12/11/back-to-basics/#:~:text=code%20uses%20the-,Shlemiel%20the%C2%A0painter%E2%80%99s%20algorithm,-.%20Who%20is%20Shlemiel

Reply:
Ahh that makes sense.

In that case, should we also have the defalt publisher enqueue images to publish, and only actually write them once the publisher is `Close`d? This implementation seems to be unique to the tarball publisher, but it requires changes to all the other implementations to account for it.

How does `ko resolve --tarball` even work? What ends up in the YAML? 🤔

Reply:
> This implementation seems to be unique to the tarball publisher

Tarball is unique as a publisher in that it's impossible to do incremental writes to it because we have to serialize everything :/

> What ends up in the YAML?

What you'd expect.

Another option here would be to construct the tarball contents in a tempdir and tar it up once we're done, but we'd still need to know when we're done :/